lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,307 @@
+
+  static void _recursiveGetExpressions(ConditionScope conditionScope,
+      List<Expression> expressions, Iterable<Element> elements, bool value) {
+    if (conditionScope == null) {
+      return;
+    }
+    for (final element in conditionScope.environment.reversed) {
+      if (element.haveToStop(elements)) {
+        return;
+      }
+      if (element is _ConditionExpression && element.value == value) {
+        expressions.add(element.expression);
+      }
+    }
+    _recursiveGetExpressions(
+        conditionScope.outer, expressions, elements, value);
+  }
+}
+
+abstract class ConditionScopeVisitor extends UnifyingAstVisitor {
+  ConditionScope outerScope;
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
+    node.visitChildren(this);
+  }
+
+  @override
+  visitBlockFunctionBody(BlockFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    _addScope();
+    visitCondition(node.condition);
+    node.visitChildren(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+  }
+
+  @override
+  visitEmptyFunctionBody(EmptyFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  @override
+  visitExpressionFunctionBody(ExpressionFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    _addScope();
+    _addTrueCondition(node.condition);
+    node.variables?.accept(this);
+    node.initialization?.accept(this);
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addTrueCondition(node.condition);
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addFalseCondition(node.condition);
+    }
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    final elseScope = _visitElseStatement(node.elseStatement, node.condition);
+    _visitIfStatement(node);
+    _propagateUndefinedExpressions(elseScope);
+    if (_isLastStatementAnExitStatement(node.thenStatement)) {
+      _addFalseCondition(node.condition);
```

Comment:
I don't understand the case you're thinking of. Can you give me an example involving a label after an if statement?

That said, I do think that we might want to forget everything we think we know when we hit a label because we don't know what the state of any variable will be when execution reaches the break or continue that will transfer control to that label. Unfortunately, the same is true for any do, for, switch or while statement because they all have an implicit empty label.